### PR TITLE
docstore: implement update queries

### PR DESCRIPTION
**Pull request by jba**

### CODE REVIEW

**Comment by jba** [comment]
@shantuo ping

File: internal/docstore/firedocstore/fs.go [hunk 1]
```diff
@@ -634,23 +648,37 @@ func toServiceFieldPathComponent(key string) string {
 // revisionPrecondition returns a Firestore precondition that asserts that the stored document's
 // revision matches the revision of doc.
 func revisionPrecondition(doc driver.Document) (*pb.Precondition, error) {
+	rev, err := revisionTimestamp(doc)
```
Comment (shantuo) [inline internal/docstore/firedocstore/fs.go]: Why create these two functions? The 3 names look similar and may cause some confusion and it seems you can use the `revisionPrecondition` in the above scenarios. At least one of the three can be saved.

File: internal/docstore/firedocstore/query.go [hunk 2]
```diff
@@ -357,12 +357,32 @@ func (c *collection) QueryPlan(q *driver.Query) (string, error) {
...
+}
+
 // For delete and update queries, limit the number of write actions per RPC, to bound
 // client memory.
 // This is a variable so it can be modified for tests.
 var maxWritesPerRPC = 500
 
-func (c *collection) RunDeleteQuery(ctx context.Context, q *driver.Query) error {
+// runActionQuery runs the query, calls writes for each returned document, and then commits those writes.
+func (c *collection) runActionQuery(ctx context.Context, q *driver.Query, writes func(*pb.Document) ([]*pb.Write, error)) error {
```
Comment (shantuo) [inline internal/docstore/firedocstore/query.go]: nit: s/runActionQuery/runWriteQuery, `RunGetQuery` doesn't use it.

File: internal/docstore/firedocstore/fs.go (hunk 1 above)
Comment (jba) [reply internal/docstore/firedocstore/fs.go]: I use preconditionFromTimestamp in newUpdateWrites and revisionTimestamp in updateWrites.

File: internal/docstore/firedocstore/query.go (hunk 2 above)
Comment (jba) [reply internal/docstore/firedocstore/query.go]: Done.

**Comment by jba** [comment]
> Also revert the change of dynamo record tests since they are not changed.

I tried to do this. On my branch the files in master are the same. I don't know why they show up as diffs here.

**Pull request merged**